src/ConnectedMode/ServerSentEvents/ServerSentEventPump.cs

```diff
@@ -0,0 +1,82 @@
+﻿/*
+ * SonarLint for Visual Studio
+ * Copyright (C) 2016-2023 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using SonarLint.VisualStudio.Core.ServerSentEvents.Issues;
+using SonarLint.VisualStudio.Core.ServerSentEvents.TaintVulnerabilities;
+using SonarQube.Client.Models.ServerSentEvents;
+using SonarQube.Client.Models.ServerSentEvents.ClientContract;
+using System;
+using System.ComponentModel.Composition;
+using System.Threading.Tasks;
+
+namespace SonarLint.VisualStudio.ConnectedMode.ServerSentEvents
+{
+    internal interface IServerSentEventPump
+    {
+        Task PumpAllAsync(IServerSentEventsSession session,
+            IIssueChangedServerEventSourcePublisher issueChangedServerEventSourcePublisher,
+            ITaintServerEventSourcePublisher taintServerEventSourcePublisher);
+    }
+
+    [Export(typeof(IServerSentEventPump))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    internal class ServerSentEventPump : IServerSentEventPump
+    {
+        private readonly IServerSentEventsFilter serverSentEventsFilter;
+
+        [ImportingConstructor]
+        public ServerSentEventPump(IServerSentEventsFilter serverSentEventsFilter)
+        {
+            this.serverSentEventsFilter = serverSentEventsFilter;
+        }
+
+        public async Task PumpAllAsync(IServerSentEventsSession session,
```

Comment:
Why doesn't it receive the publishers in the CTOR? they are always the same.

Reply:
it's a bit of a tricky situation. the channels are disposables, but the pump isn't. I decided to limit the ownership scope to reduce the confusion, but maybe I didn't achieve it

Reply:
The channels are not disposable though, they're singletons. We just no longer pump data into them if the connect ended.

Reply:
But they do implement IDisposable and throw after the disposal. It only happens on shutdown, but still.

Reply:
moved to ctor